Add headless auth API for native mobile apps

rabble opened the pull request:
## Summary

Adds pure JSON authentication endpoints for native mobile apps (Flutter, React Native, etc.) that want to handle their own UI instead of using web views.

## New Endpoints

- `POST /api/headless/register` - Register new user, returns `device_code` for email verification polling
- `POST /api/headless/login` - Login existing user, returns authorization code directly  
- `POST /api/headless/authorize` - Create new authorization for already-authenticated user

## Flow

**Registration:**
1. App calls `/api/headless/register` with email/password
2. Gets back `device_code` and shows "check your email" 
3. Polls `/api/oauth/poll?device_code=xxx` until email verified
4. Exchanges code for `bunker_url` via `/api/oauth/token`

**Login:**
1. App calls `/api/headless/login` with email/password
2. Gets back authorization `code` directly
3. Exchanges code for `bunker_url` via `/api/oauth/token`

## Files Changed

- `api/src/api/http/headless.rs` - New module with handlers
- `api/src/api/http/mod.rs` - Module declaration
- `api/src/api/http/routes.rs` - Route registration
- `api/tests/headless_auth_test.rs` - Integration tests (6 tests, all pass)
- `docs/HEADLESS_AUTH.md` - Documentation with Flutter examples

## Testing

All 6 new tests pass. No cookies, no web views, any origin allowed.

rabble committed:
Add headless auth API for native mobile apps

Adds pure JSON endpoints for Flutter/native apps to handle their own UI:
- POST /api/headless/register - returns device_code for email polling
- POST /api/headless/login - returns auth code directly
- POST /api/headless/authorize - creates new auth for existing session

No web views, no cookies. Any app can call these endpoints.

rabble committed:
Fix formatting

dcadenas commented:
I have significant concerns about this approach, some strategic, some technical.

  **Security concerns**

  The immediate issue is permissive CORS on these endpoints - any website can call `/api/headless/login` and harvest credentials directly.

  This also implements OAuth 2.0 ROPC (Resource Owner Password Credentials), which I intentionally removed. [RFC 9700](https://datatracker.ietf.org/doc/rfc9700/) (January 2025) states: **"The resource owner password credentials grant MUST NOT be used."** [OAuth 2.1](https://oauth.net/2.1/) removes it entirely. While there's [debate about first-party apps](https://blog.logto.io/deprecated-ropc-grant-type), the standards don't carve out an exception. ROPC is a single request/response, so challenge-based authentication is impossible - [it's incompatible with MFA](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth-ropc) and passkeys out of the box. You could build those on top, but then you're implementing authentication features in the app instead of centralizing them at login.divine.video - duplicating work we'll need to do (or already have) in the provider anyway.

  **What the industry does**

  Google, GitHub, and ChatGPT all use system browser auth for their own first-party apps:

  Google:
  <img src="https://github.com/user-attachments/assets/43bc8e53-7ec8-4efd-88a7-6578ed50cd83" width="200px" />

  ChatGPT:
  <img src="https://github.com/user-attachments/assets/60c8d1db-d02b-4831-9023-84d96e14f115" width="200px" />

  Github:
  <img src="https://github.com/user-attachments/assets/caab6bb0-b194-45c6-8401-a4a01ecd3d0e" width="200px" />

  Looking at federated social networks specifically: [Mastodon removed password grant entirely in version 4.4](https://docs.joinmastodon.org/spec/oauth/). Bluesky's app still uses ROPC, but their official OAuth announcement states the app "has not **yet** been updated to use OAuth" - that "yet" signals intent. [Blacksky is actively implementing OAuth](https://github.com/blacksky-algorithms/rsky/issues/60), and the [recommended OAuth implementation for AT Protocol](https://github.com/graze-social/aip) doesn't include password grant at all.

  **Why this matters for us**

  We're building an identity provider for the Nostr ecosystem. Third-party apps shouldn't collect passwords directly - that's the whole point of [OAuth's delegated authorization model](https://developer.okta.com/docs/concepts/oauth-openid/). If Divine uses a different flow than what we'd recommend for third-parties, we're not training users on the pattern we want them to expect: trust login.divine.video, just like users trust accounts.google.com or github.com.

  With headless auth, users never see the domain. The password field in Divine looks identical to a password field in a phishing app that anyone can implement. If we train users to trust login.divine.video from the start, we're in a much better position to help Nostr adoption - new non-technical users get a familiar, secure login experience across the ecosystem. We have the opportunity to start right rather than migrate later - and we already have a server implementation doing exactly this.

  **On "web views"**

  The PR describes this as avoiding "web views," but if that means avoiding web pages entirely, that misses the core security model of OAuth. The redirect to a centralized login page IS the point - users provide their password to the identity provider, not to the app. The current flow uses ASWebAuthenticationSession (iOS) and Chrome Custom Tabs (Android) - system browser sheets, not embedded web views. This can be fully Divine-branded, just like Google/GitHub/ChatGPT brand their OAuth pages.

  We could discuss what specific UX friction we're trying to solve here.

rabble commented:
Yah na... @dcadenas you don't understand the primary goal of keycast for me and divine. The primary goal is a FIRST PARTY auth system and that it's seemlessly integrated in to divine. The goal 10000% is to never have to train users on anything. First and foremost what we're doing is providing a login system to divine, for divine users, from divine. Anything else like oauth to other nostr services or ability to get an nsec bunker is cool, supporting an open ecosystem, definitely something we want, but not the goal or purpose of keycast for divine. 

I have zero desire to support the corruption and broken newer versions of OAUTH. It was a good standards at 1.0, broken but workable at 2.0... beyond that it became corporate slop that was designed for big companies like OKTA and google and not workable for an open ecosystem. 

So i'm fine with adding CORS protection back, doesn't matter from the mobile app. We are not making users from divine, login with divine... doesn't make sense. It's our app, our auth server, we control everything, we ask suers for username and password directly in the app and do not throw up a web UI.. this is not negotiable.

dcadenas committed:
fix: restrict headless auth CORS to first-party origins

Headless endpoints accept passwords directly, so restrict CORS to
ALLOWED_ORIGINS (auth_cors) instead of allowing any origin (public_cors).
Also add divine.video to allowed origins in cloudbuild.yaml.
